evennia/typeclasses/attributes.py

```diff
@@ -414,6 +470,366 @@ def reset_cache(self):
+        raise NotImplementedError()
+
+    def batch_add_update(self, attr_obj, category, lock_storage, new_value):
+        """
+        Called speciallly by batch add. For the database backend, this is a method
+        of updating that can alter category and lock-storage. Is simply an 'update' for In-Memory.
+        
+        Args:
+            attr_obj (IAttribute): The Attribute being altered.
+            category (str or None): The attribute's (new) category.
+            lock_storage (str): The attribute's new locks.
+            new_value (obj): The Attribute's new value.
+
+        """
+        raise NotImplementedError()
+
+    def batch_add_create(self, keystr, category, new_value, lockstring):
+        raise NotImplementedError()
+
+    def batch_add_finish(self, attr_objs):
+        raise NotImplementedError()
+
+    def batch_add(self, *args, **kwargs):
+        """
+        Batch-version of `add()`. This is more efficient than
+        repeat-calling add when having many Attributes to add.
+
+        Args:
+            indata (list): List of tuples of varying length representing the
+                Attribute to add to this object. Supported tuples are
+                    - `(key, value)`
+                    - `(key, value, category)`
+                    - `(key, value, category, lockstring)`
+                    - `(key, value, category, lockstring, default_access)`
+
+        Raises:
+            RuntimeError: If trying to pass a non-iterable as argument.
+
+        Notes:
+            The indata tuple order matters, so if you want a lockstring
+            but no category, set the category to `None`. This method
+            does not have the ability to check editing permissions like
+            normal .add does, and is mainly used internally. It does not
+            use the normal self.add but apply the Attributes directly
+            to the database.
+
+        """
+        new_attrobjs = []
+        for tup in args:
+            if not is_iter(tup) or len(tup) < 2:
+                raise RuntimeError("batch_add requires iterables as arguments (got %r)." % tup)
+            ntup = len(tup)
+            keystr = str(tup[0]).strip().lower()
+            new_value = tup[1]
+            category = str(tup[2]).strip().lower() if ntup > 2 and tup[2] is not None else None
+            lockstring = tup[3] if ntup > 3 else ""
+
+            attr_objs = self.get_cache(keystr, category)
+
+            if attr_objs:
+                attr_obj = attr_objs[0]
+                # update an existing attribute object
+                self.batch_add_update(attr_obj, category, lockstring or "", new_value)
+            else:
+                new_attr = self.batch_add_create(keystr, category, new_value, lockstring)
+                new_attrobjs.append(new_attr)
+                self.set_cache(keystr, category, new_attr)
+        if new_attrobjs:
+            self.batch_add_finish(new_attrobjs)
+
+    def delete_attr(self, attr):
```

Comment:
Consistency is important. Decide in your naming if you should spell out `attribute/s` or not. So either this should be `delete_attribute` or the next method should `delete_attrs`. 
Overall, I think naming all these method as with the full `_attribute(s)` is likely clearer.